pattern/src/lib.rs

```diff
@@ -1,9 +1,107 @@
+//!   the value captured by the pattern is discarded.
+//!
+//!   Each pattern captures exactly one value. Often, it is the [`Net`] or
+//!   [`Value`] it is being matched against. In rare cases, the captured value
+//!   is of the unit type `()`, in which case it is usually discarded by the
+//!   user.
+//! - a number of expression arguments, e.g. `("cin")`. Expression arguments
+//!   are evaluated according to normal Rust rules, and the results are passed
+//!   to the pattern. Each pattern has a fixed number of expression arguments
+//!   it expects.
+//!
+//!   If multiple expression arguments are being passed to the pattern, each
+//!   requires its own pair of parentheses, e.g. `PFeline@cat ("meow") (42)`
+//! - a number of pattern arguments, e.g. `[PZero]`. Pattern arguments
+//!   follow the pattern call syntax being described.
+//!
+//!   Each pattern has a fixed number of pattern arguments it expects.
+//!   Most commonly, they correspond to the inputs of the cell being matched.
+//!
+//!   If you do not wish to perform further matching on the inputs, use `[PAny]`.
+//!
+//! Much like `match`, [`netlist_matches!`] supports `if` and `if let` guards.
+//! For example:
+//!
+//! ```no_run
+//! # use prjunnamed_netlist::{Design, Net, Value};
+//! # use prjunnamed_pattern::netlist_matches;
+//! # use prjunnamed_pattern::patterns::*;
+//! let matcher = netlist_matches! {
+//!     [PXor [PAny@a] [PAny@b]] if a == b => Value::zero(a.len());
+//! };
+//! # let design: Design = todo!();
+//! # let target: Value = todo!();
+//! # let result: Option<Value> = matcher(&design, &target);
+//! ```
+
+/// A pattern that can be matched against a `Target`.
+///
+/// Most commonly, `Target` will be [`Value`] or [`Net`], but it could also be
+/// something like `u32`.
+///
+/// A pattern must also have a `new` function that will be called by the macro
+/// to construct an instance of the pattern. The arguments of the `new` function
+/// consist of all the expression arguments, followed by all the pattern
+/// arguments.
 pub trait Pattern<Target> {
+    /// The capture type resulting from matching this pattern *and all the
+    /// patterns passed as arguments*.
+    ///
+    /// The macros expect this to be a tuple, where the first element is
+    /// the value captured by the pattern itself, and the following elements
+    /// correspond to the `Capture`s of all the child patterns.
     type Capture;
 
     fn execute(&self, design: &dyn DesignDyn, target: &Target) -> Option<Self::Capture>;
 }
 
+/// Matches a [`Value`] against a list of patterns and runs the body
+/// corresponding to the first matching pattern.
+///
+/// To be specific, `netlist_matches!` returns an
+/// `impl Fn(&dyn DesignDyn, &Value) -> Option<T>`. If no pattern matches
+/// the value provided, the function will return `None`. Otherwise, it will
+/// call `.into()` on the result of the matching arm and return `Some`.
```

Comment:
Worth saying that all of the rules will be evaluated top-down in the order they are written up until a rule matches, since this has a major impact on semantics. This *does* work like the ML-style pattern matching but like I said above I think it's worth describing it not as a diff.